### ISSUE TITLE
Alpaca dataset padding 0 for tokens and -1 rather than -100 for labels

### ISSUE THREAD

**Issue by wang-sj16** [facebookresearch/llama-recipes#74]
### System Info

In the alpaca dataset, the labels use 0 as padded tokens rather than `tokenizer.pad_token_id (32000)`. For labels, it also uses -1 rather than -100 to indicate unrelated labels. The incorrect padding will cause broken output after fine-tuning. 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

```python
padding = self.max_words - example.shape[0]
if padding > 0:
    example = torch.cat((example, torch.zeros(padding, dtype=torch.int64) - 1))
elif padding < 0:
    example = example[: self.max_words]
labels = copy.deepcopy(example)
labels[: len(prompt)] = -1
```

### Error logs

After fine-tuning with the incorrect paddings, the model output meaningless codes rather than wanted results.

### Expected behavior

I change the padding methods with the following codes and got normal behavior:
replace `alpaca_dataset.py` line 58-71 to:

```python
        labels = copy.deepcopy(example)
        labels[: len(prompt)] = IGNORE_INDEX
        padding = self.max_words - example.shape[0]
        if padding > 0:
            example = torch.cat((example, self.tokenizer.pad_token_id*torch.zeros(padding, dtype=torch.int64)))
            labels = torch.cat((labels, IGNORE_INDEX*torch.ones(padding, dtype=torch.int64)))
        elif padding < 0:
            example = example[: self.max_words]
            labels = labels[: self.max_words]

        example_mask = example.ne(self.tokenizer.pad_token_id)
        example_mask = example_mask.float()
```
where `IGNORE_INDEX=-100` 

<img width="2019" alt="image" src="https://github.com/facebookresearch/llama-recipes/assets/28758180/0e55003c-6ca8-436e-a820-5096ae222474">

**Comment by HamidShojanazeri** [comment]
@wang-sj16 thanks for opening this issue, would you be interested to send a PR?

**Comment by wang-sj16** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    Thank you for the reply, will send a PR tomorrow!
    
          
        
            
    
        font{
            line-height: 1.6;
        }
    

    
        font{
            line-height: 1.6;
        }
    
Shijie WangDepartment of Computer Science, Brown ***@***.******@***.***


        
    
     
    

  
    ---- Replied Message ----
  
  
    
      
        
         From 
        
        
            Hamid ***@***.***>
            
        
      
      
        
         Date 
        
        
        07/31/2023 21:50
        
      
      
        
         To 
        
        
         
          
            ***@***.***>
            
          
        
      
      
        
         Cc 
        
        
          
            Shijie ***@***.***>
            ,
          
          
            ***@***.***>
            
          
        
      
      
        
         Subject 
        
        
              Re: [facebookresearch/llama-recipes] Alpaca dataset padding 0 for tokens and -1 rather than -100 for labels (Issue #74)
        
      
    
  
  
@wang-sj16 thanks for opening this issue, would you be interested to send a PR?

**Comment by HamidShojanazeri** [comment]
@wang-sj16 Thanks would be great to have your PR.

**Comment by HamidShojanazeri** [comment]
Just note, we have added the [<PAD> token as the special token to tokenizer](https://github.com/facebookresearch/llama-recipes/blob/main/llama_finetuning.py#L111). This by default is `None` in [HF tokenizer](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/tokenization_llama.py#L109). We can use the tokenizer for padding just we would need to resize the embedding as `model.resize_token_embeddings(model.config.vocab_size + 1)`.

**Comment by Luobots** [comment]
The same bug observed! 😂
I want to know wether the original script can influence the model?
For example, `example` use `0` as the `pad_token_id`, should it bring bad influence?
I think it won't because the `attention_mask` avoids them?

**Comment by wang-sj16** [comment]
> The same bug observed! 😂 I want to know wether the original script can influence the model? For example, `example` use `0` as the `pad_token_id`, should it bring bad influence? I think it won't because the `attention_mask` avoids them?

I trained with the original script and during generation it just output gibberish. I think the main influence is from the incorrect label padding that makes the model learn from non-sense 0 labels

**Comment by Luobots** [comment]
> I trained with the original script and during generation it just output gibberish. I think the main influence is from the incorrect label padding that makes the model learn from non-sense 0 labels

Thanks for your reply!
Yes I found the wrong behavior too. I just set the `pad_token_id` to -100 like you too.
However, I can't figure out the function of codes here [add pad token here](https://github.com/facebookresearch/llama-recipes/blob/1387b76e110970f26c976bec9d43f9da961470f1/llama_finetuning.py#L111) 🥹.

**Comment by HamidShojanazeri** [comment]
addressed in this[ PR](https://github.com/facebookresearch/llama-recipes/pull/107).

**Comment by farris** [comment]
This fixed a lot of existing issues for me... thanks 👍
